src/main/java/com/avispl/dal/communicator/polycom/videoos/PolycomVideoOS.java

```diff
@@ -0,0 +1,1046 @@
+    }
+
+    /**
+     * Check if the device adapter is initialized and run internalInit() if not.
+     *
+     * @throws Exception if any error occurs
+     */
+    private void checkAndInitialize() throws Exception {
+        if (!isInitialized()) {
+            internalInit();
+        }
+    }
+
+    /**
+     * Generate CallStatus instance based on callId and callStatusState parameters
+     *
+     * @param callId id of the conference
+     * @param callStatusState state of the call to use
+     * @return {@link CallStatus} instance, indicating the requested status of the call
+     */
+    private CallStatus generateCallStatus(String callId, CallStatus.CallStatusState callStatusState){
+        CallStatus callStatus = new CallStatus();
+        callStatus.setCallId(callId);
+        callStatus.setCallStatusState(callStatusState);
+        return callStatus;
+    }
+
+    /**
+     * Check if outcoming video feed is muted
+     *
+     * @return boolean outcoming video feed mute status
+     * @throws Exception during http communication
+     */
+    private Boolean retrieveVideoMuteStatus() throws Exception {
+        JsonNode response = doGet(VIDEO_MUTE, JsonNode.class);
+        return getJsonProperty(response, "result", Boolean.class);
+    }
+
+    /**
+     * Set local video feed mute status
+     *
+     * @param status boolean indicating the target outcoming video feed state
+     * @return boolean value, indicating the success of the mute status change.
+     * @throws Exception during http communication
+     */
+    private Boolean updateVideoMuteStatus(boolean status) throws Exception {
+        ObjectNode request = JsonNodeFactory.instance.objectNode();
+        request.put("mute", status);
+        JsonNode response = doPost(VIDEO_MUTE, request, JsonNode.class);
+        return getJsonProperty(response, "success", Boolean.class);
+    }
+
+    /**
+     * Get media statistics related to the current conference call
+     *
+     * @param conferenceCallMediaStats data retrieved from the VidoOS API
+     * @param audioChannelStats        entity containing audio stats
+     * @param videoChannelStats        entity containing video stats
+     * @param callStats                entity containing call stats
+     */
+    private void retrieveConferenceCallMediaStats(ArrayNode conferenceCallMediaStats, AudioChannelStats audioChannelStats,
+                                                  VideoChannelStats videoChannelStats, CallStats callStats) {
+        conferenceCallMediaStats.forEach(jsonNode -> {
+            switch (Objects.requireNonNull(getJsonProperty(jsonNode, "mediaDirection", String.class))) {
+                case "RX":
+                    switch (Objects.requireNonNull(getJsonProperty(jsonNode, "mediaType", String.class))) {
+                        case "AUDIO":
+                            audioChannelStats.setBitRateRx(getJsonProperty(jsonNode, "actualBitRate", Integer.class));

```

Comment:
Maksym - I'm not sure why you are parsing JsonNodes manually in pretty much all adapter implementations. It makes code much longer and open it for bugs. Unless there are valid reasons for that (e.g. API responses are not standardized), there is no need to do it! The proper way is to create DTO hierarchy for response data and let REST template know what object you expect and it will do all the parsing for you.
I'm not asking to change it for this adapter, but, in the future if I see using JsonNode directly I'll have to ask for justification.

Reply:
Ok, thank you. Will make sure DTOs are used in future.